Objects/longobject.c

```diff
@@ -3925,6 +4081,12 @@ l_divmod(PyLongObject *v, PyLongObject *w,
         }
         return 0;
     }
+#if WITH_PYLONG_MODULE
+    if (Py_ABS(Py_SIZE(w)) > 1000) {
```

Comment:
Maybe my setup is iffy, but I'm seeing this some things slow down 10x. I ran this code mostly copied from pyperformance's bm_pidigits:

<details>

```python
import itertools
from time import perf_counter
icount = itertools.count
islice = itertools.islice

def gen_x():
    return map(lambda k: (k, 4 * k + 2, 0, 2 * k + 1), icount(1))


def compose(a, b):
    aq, ar, as_, at = a
    bq, br, bs, bt = b
    return (aq * bq,
            aq * br + ar * bt,
            as_ * bq + at * bs,
            as_ * br + at * bt)


def extract(z, j):
    q, r, s, t = z
    return (q * j + r) // (s * j + t)


def gen_pi_digits():
    z = (1, 0, 0, 1)
    x = gen_x()
    while 1:
        y = extract(z, 3)
        while y != extract(z, 4):
            z = compose(z, next(x))
            y = extract(z, 3)
        z = compose((10, -10 * y, 0, 1), z)
        yield y


def calc_ndigits(n):
    return list(islice(gen_pi_digits(), n))


for n in 100, 200, 500, 1_000, 2_000, 5_000, 10_000, 20_000:
    loops = 1
    while True:
        t0 = perf_counter()
        for i in range(loops):
            calc_ndigits(n)
        t1 = perf_counter()
        dt = t1 - t0
        if dt > 1.0:
            break
        else:
            loops *= 2
    print(f"{n:_}\t{dt/loops:.03}")
```

</details>

```
Before this PR:
100     0.000808
200     0.00274
500     0.0138
1_000   0.0513
2_000   0.207
5_000   1.37
10_000  5.76
20_000  26.5

After this PR:
100     0.000806
200     0.00278
500     0.0142
1_000   0.182
2_000   1.5
5_000   11.1
10_000  49.9
20_000  2.27e+02
```

Reply:
I don't think there's anything wrong with your setup, but instead that the new code shouldn't be called here. I believe `extract()` always returns an integer in `range(10)`. That is, the quotient it returns is always tiny, never more than 4 bits. "Schoolbook" division in linear-time in such cases. finishing after the very first multiply-and-subtract loop (given that we use 30-bit "digits" internally - we only need one such step even for quotients substantially fatter than 4 bits).

So testing for the size of the denominator alone is too crude. This is where I hope Mark D magically appears, since I never fought my way to a real understanding of the fancy algorithm, but he did 😃. I expect we certainly don't want to use it unless the quotient is at least fatter than one internal digit.

Reply:
I added a test `(size_numerator - size_denominator) > 1`.  I think that's what you mean Tim and it seems to fix the bm_pidigits benchmark.  I'm attaching a little script can be used to test.

Reply:
On the general track, but I think the order of the subtraction was backwards (and already added a suggested fix). Either way, I believe `bm_digits` will never invoke the Python division now.

A "good" cutoff remains to be determined. I can't guess whether the difference in internal-digit count "should` be >= 1, >=2, >= 3, ... only that < 1 is bound to lose. That requires empirical testing.

One way is to try three-argument `pow()` with bigger and bigger modulus, but with randomized arguments and "largish" exponents. Where `m = modulus.bit_length()`, that will tend to generate `2m`-bits divided by `m`- bits cases internally. The number of such cases an instance generates is around `the_exponent.bit_length()`.

BTW, I'm sure you'll be delighted to hear that GMP has Yet Another int division algorithm solely devoted to "small quotient" cases :wink".

Reply:
Yeah, I got 'v' and 'w' confused and swapped.  I fixed that but after making another benchmark, it seems the cutoff test is still not very good (i.e. not conservative enough).  Maybe a better test is that the numerator is N digits larger than denominator?

```
import timeit
import random
import string
import _pylong

def rand_digits(n):
    return random.randint(10**n, 10**(n+1))

def main():
    print('Note: "time" is for py version, "ratio" is py/builtin time.')
    print(f'{"n-digits":>8} {"d-digits":>8} {"time [s]":>8} {"ratio":>8}')
    for digits in [2_000, 5_000, 10_000, 50_000]:
        for diff_exp in [0, 1, 2, 3, 4, 5]:
            digits_denom = digits - 10**diff_exp
            if digits_denom <= 1:
                continue
            # Generate list of random arguments with specific number of digits
            args = []
            for i in range(1000):
                a = rand_digits(digits)
                b = rand_digits(digits_denom)
                args.append((a, b))
            times = []
            n = 1
            for my_divmod in [divmod, _pylong.int_divmod]:
                def run():
                    for a, b in args:
                        q, r = my_divmod(a, b)
                t = timeit.Timer('run()', globals=vars())
                r = t.timeit(n)
                times.append(r/n)
            #print(str(a // b)[-10:])
            base_time, py_time = times
            print(f'{digits:>8} {digits_denom:>8} {py_time:8.3f}'
                  f' {py_time/base_time:8.2f}')

if __name__ == "__main__":
    main()
```
Benchmark output
```
Note: "time" is for py version, "ratio" is py/builtin time.
n-digits d-digits time [s]    ratio
    2000     1999    0.011    13.25
    2000     1990    0.011     9.86
    2000     1900    0.014     4.13
    2000     1000    0.020     1.41
    5000     4999    0.043    20.45
    5000     4990    0.044    16.03
    5000     4900    0.046     5.69
    5000     4000    0.061     1.19
   10000     9999    0.092    22.19
   10000     9990    0.087    17.67
   10000     9900    0.090     5.63
   10000     9000    0.124     1.08
   50000    49999    0.359    19.51
   50000    49990    0.371    15.72
   50000    49900    0.413     5.22
   50000    49000    0.721     1.16
   50000    40000    1.918     0.39
```

Reply:
> Maybe a better test is that the numerator is N digits larger than denominator?

For some prior art, it looks like that's roughly what OpenJDK's `BigInteger`s do:

```python
if denom_limbs > 80 and num_limbs - denom_limbs > 40:
    do_BurnikelZiegler(...)
```

It seems like we need some wildly different constants though.

Reply:
Thanks for looking that up! That's the right direction, and shouldn't be too surprising. The division code here is much more "fiddly" and involved than the base-conversion algorithms, and Mark noted on the original many ways in which it could benefit from being coded in C. Note, e.g., that just breaking the numerator into slices each with a `denominator.bit_length()` width is a quadratic-time task in the Python code (although with a small constant factor).